### PR TITLE
Issue #KN-63 feat: Mandatory & optional configuration

### PR DESCRIPTION
**Pull request by gouravmore**
<table>
<td>
  <a href="https://project-sunbird.atlassian.net/browse/KN-63" title="KN-63" target="_blank">RFCKN-63</a>  PDF player Mandatory & optional configuration
  </td></table>
  <br />
 

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

### Type of change

Please choose appropriate options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes in the below checkboxes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Ran Test A
- [ ] Ran Test B

**Test Configuration**:
* Software versions:
* Hardware versions:

### Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -127,34 +127,55 @@ var playerConfig = {
...
+|Property Name| Description| Default Value |  Mandatory/Optional
+|--|----------------------|--| --|
+| `config` | It is an `object` it contains the `sideMenu`, these will be used to configure the canvas  | ```{  sideMenu: {"showShare": true, "showDownload": true, "showReplay": true, "showExit": false,"showPrint": true}}``` | Mandatory |
+| `config.sideMenu.showShare` | It is  `boolean` to show/hide share button in side menu| ```true```| Optional |
+| `config.sideMenu.showDownload` | It is  `boolean` to show/hide download button in side menu| ```true```| Optional |
+| `config.sideMenu.showReplay` | It is  `boolean` to show/hide replay button in side menu| ```true```| Optional |
+| `config.sideMenu.showExit` | It is  `boolean` to show/hide exit button in side menu| ```false```| Optional |
+| `config.sideMenu.showPrint` | It is  `boolean` to show/hide print button in side menu| ```true```| Optional |
+| `metadata` | It is an `object` which defines content metadata json object (from API response take -> response.result.content) | ```{}```| Mandatory |
```
Comment (vinukumar-vs) [inline README.md:178]: Put all mandatory fields on top & then add the optional fields

File: projects/sunbird-pdf-player/src/lib/playerInterfaces.ts [hunk 2]
```diff
@@ -25,15 +25,15 @@ export interface ObjectRollup {
 }
 
 export interface Context {
-    mode: string;
+    mode?: string;
```
Comment (vinukumar-vs) [inline projects/sunbird-pdf-player/src/lib/playerInterfaces.ts:28]: If all these properties are optional, then why "context" is mandatory in the "PlayerConfig"?

Comment (vinukumar-vs) [reply projects/sunbird-pdf-player/src/lib/playerInterfaces.ts:28]: If you make all are optional, How the telemetry events will generate if any required property is missing for telemetry like did, uid etc..

**Review by vinukumar-vs — CHANGES_REQUESTED** [review]

File: projects/sunbird-pdf-player/src/lib/services/viewer.service.ts [hunk 3]
```diff
@@ -46,9 +46,11 @@ export class ViewerService {
     } else {
       this.src =  metadata.streamingUrl || metadata.artifactUrl;
     }
-    if (context.userData) {
+    if (context?.userData) {
       const { userData: { firstName, lastName } } = context;
       this.userName = firstName === lastName ? firstName : `${firstName} ${lastName}`;
+    }else{
```
Comment (HarishGangula) [inline projects/sunbird-pdf-player/src/lib/services/viewer.service.ts:52]: Code indentation missing

File: README.md [hunk 4]
```diff
@@ -227,34 +227,58 @@ var playerConfig = {
...
-| `cdata` | It is an `array` which defines the correlation data | ```[]```|
+| `identifier` | It is  `string` of uniq content id | Mandatory |
+| `Name` | It is  `string` to represent the name of the content or pdf | Mandatory |
+| `artifactUrl` | It is  `string` url  to load the pdf from artifact url | Mandatory |
+| `streamingUrl` | It is  `string` url  to load the pdf from streaming url | Optional |
+| `isAvailableLocally` | It is a `boolen` value which indicate the content is locally available | Optional |
+| `basePath` | It is `string` to represent the base path of the pdf file | Optional |
+| `baseDir` | It is `string` to represent the base path of the pdf file | Optional |
+| `compatibilityLevel` | It is `number` to represent the compatibility level | Optional |
+| `pkgVersion` | It is `number` to represent the version of the current packages | Optional |
```
Comment (HarishGangula) [inline README.md:242]: What is the default value if this is optional or behaviour should be updated here and all other optional fields

**Review by HarishGangula — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (gouravmore) [reply README.md]: Updated READMD file

File: README.md [hunk 5]
```diff
@@ -227,34 +227,79 @@ var playerConfig = {
...
+  baseDir?: string;
+}
+  ```
+  
+ In metadata the following properties are mandatory to play the content.
+ 
+ |Property Name| Description|  Mandatory/Optional| Without field | Comment
+|--|----------------------|--| --| --|
+| `identifier` | It is  `string` of uniq content id | Mandatory | Unable to load the content error | Its a unique content id so Its a required to log the telemetry and other data against contnet|
+| `Name` | It is  `string` to represent the name of the content or pdf | Mandatory | Unable to load the content error | Its a required to show the name of the pdf while loading the pdf|
```
Comment (HarishGangula) [inline README.md:255]: keep **N** as small in **Name**

File: README.md [hunk 6]
```diff
@@ -227,34 +227,79 @@ var playerConfig = {
...
+| `artifactUrl` | It is  `string` url  to load the pdf from artifact url | Mandatory | Unable to load the content error | It is required to load the pdf file|
+| `streamingUrl` | It is  `string` url  to load the pdf from streaming url | Optional | Unable to load the content error | It is optional field. This is required if you want to load the streaming pdf url|
+| `isAvailableLocally` | It is a `boolen` value which indicate the content is locally available | Optional | Content will not load offline | It is required to know - the content is downloaded and can be play offline|
```
Comment (HarishGangula) [inline README.md:258]: fix spelling mistake **boolean**

File: README.md [hunk 7]
```diff
@@ -227,34 +227,79 @@ var playerConfig = {
...
+| `basePath` | It is `string` to represent the base path of the pdf file | Optional | Content will not load offline | It is required to load the pdf file in offline use case|
+| `baseDir` | It is `string` to represent the base path of the pdf file | Optional | Content will not load offline |  It is required to load the pdf file in offline use case |
+| `compatibilityLevel` | It is `number` to represent the compatibility level | Optional | Defalut compatibilityLevel 4 will set | Its an Optional field
+| `pkgVersion` | It is `number` to represent the version of the current packages | Optional | Defalut compatibilityLevel 1.0 will set | its an optional field
```
Comment (HarishGangula) [inline README.md:262]: Here you need to mention the Default pkgVersion

File: README.md [hunk 8]
```diff
@@ -227,34 +227,79 @@ var playerConfig = {
...
+| `artifactUrl` | It is  `string` url  to load the pdf from artifact url | Mandatory | Unable to load the content error | It is required to load the pdf file|
+| `streamingUrl` | It is  `string` url  to load the pdf from streaming url | Optional | Unable to load the content error | It is optional field. This is required if you want to load the streaming pdf url|
+| `isAvailableLocally` | It is a `boolen` value which indicate the content is locally available | Optional | Content will not load offline | It is required to know - the content is downloaded and can be play offline|
+| `basePath` | It is `string` to represent the base path of the pdf file | Optional | Content will not load offline | It is required to load the pdf file in offline use case|
+| `baseDir` | It is `string` to represent the base path of the pdf file | Optional | Content will not load offline |  It is required to load the pdf file in offline use case |
+| `compatibilityLevel` | It is `number` to represent the compatibility level | Optional | Defalut compatibilityLevel 4 will set | Its an Optional field
```
Comment (HarishGangula) [inline README.md:261]: Default spelling mistake

File: README.md [hunk 9]
```diff
@@ -227,34 +227,79 @@ var playerConfig = {
...
-| `pdata` | It is an `object` which defines the producer information it should have identifier and version and canvas will log in the telemetry| ```{'id':'in.sunbird', 'ver':'1.0'}```|
-| `contextRollup` | It is an `object` which defines content roll up data | ```{}```|
-| `tags` | It is an `array` which defines the tag data | ```[]```|
-| `objectRollup` | It is an `object` which defines object rollup data | ```{}```|
-| `host` | It is  `string` which defines the from which domain content should be load|```window.location.origin```  |
-| `userData` | It is an `object` which defines user data | ```{}```|
-| `cdata` | It is an `array` which defines the correlation data | ```[]```|
+|Property Name| Description| Default Value | Mandatory/Optional| Comment|
+|--|----------------------|--|--| --|
+| `channel` | It is `string` which defines channel identifier to know which channel is currently using.| `in.sunbird` |Mandatory| Default can be set from player side.  Telemetry sdk will set by default ```“in.ekstep"```|
```
Comment (HarishGangula) [inline README.md:278]: Why user(Adaptor) need to know what player sets and what telemetry SDK, you mention what is the default value if not set

File: README.md [hunk 10]
```diff
@@ -227,34 +227,79 @@ var playerConfig = {
...
+| `env` | It is an string containing Unique environment where the event has occurred | ```"contentplayer"```|Optional| Content player set it by default to “contentplayer"|
+| `pdata` | It is an `object` which defines the producer information it should have identifier and version and canvas will log in the telemetry| ```{'id':'in.sunbird', 'ver':'1.0'}```|Mandatory| Default can be set from player side.  Telemetry sdk will set by default ```{ id: "in.ekstep", ver: "1.0", pid: "" }```|
```
Comment (HarishGangula) [inline README.md:280]: Why user(Adaptor) need to know what player sets and what telemetry SDK, you mention what is the default value if not set

File: README.md [hunk 11]
```diff
@@ -227,34 +227,79 @@ var playerConfig = {
...
+| `mode` | It is  `string` to identify preview used by the user to play/edit/preview | ```play```|Optional| Telemetry sdk will set by default as ```"play"```|
```
Comment (HarishGangula) [inline README.md:281]: For all the config what is the default value if it is optional mention that, don't confuse with telemetry SDK

File: projects/sunbird-pdf-player/src/lib/sunbird-pdf-player.service.ts [hunk 12]
```diff
@@ -22,6 +22,14 @@ export class SunbirdPdfPlayerService {
     this.config = config;
     this.playSessionId = this.utilService.uniqueId();
 
+    if (!context?.channel){
```
Comment (HarishGangula) [inline projects/sunbird-pdf-player/src/lib/sunbird-pdf-player.service.ts:25]: Where are the default configs mentioned in documentation set here

**Review by HarishGangula — CHANGES_REQUESTED** [review]

File: projects/sunbird-pdf-player/src/lib/sunbird-pdf-player.service.ts (hunk 12 above)
Comment (gouravmore) [reply projects/sunbird-pdf-player/src/lib/sunbird-pdf-player.service.ts]: Default config added if channel and pdata is not available in context

**Review by HarishGangula — APPROVED** [review]

**Pull request merged**